sds_data_manager/lambda_code/IAlirtCode/ialirt_ingest.py

```diff
@@ -421,6 +428,122 @@ def insert_data(data: list[dict], algorithm_table, instrument: str):
         logger.info(f"Inserted {instrument.upper()}.")
 
 
+def reformat_data(data):
+    """Reformat science and housekeeping data.
+
+    Parameters
+    ----------
+    data : list[dict]
+        Data product.
+
+    Returns
+    -------
+    science_data : list[dict]
+        Reformatted science data.
+    hk_data : list[dict]
+        Reformatted housekeeping data.
+    """
+    # Reformat data (remove all keep/exclude keys except hk
+    # once imap_processing is updated)
+    exclude_keys = {"apid", "met", "mag_hk_status"}
+    rename_map = {"met_in_utc": "time_utc"}
+    science_data = [
+        {rename_map.get(k, k): v for k, v in item.items() if k not in exclude_keys}
+        for item in data
+    ]
+    keep_keys = {"met_in_utc", "instrument", "mag_hk_status"}
+    hk_data = [
+        {
+            rename_map.get(k, k): ("mag_hk" if k == "instrument" and v == "mag" else v)
+            for k, v in item.items()
+            if k in keep_keys
+        }
+        for item in data
+        if item.get("instrument") == "mag"
+    ]
+
+    return science_data, hk_data
+
+
+def insert_formatted_data(
+    data: list[dict],
+    data_table,
+    instrument: str,
+):
+    """Insert database rows.
+
+    Parameters
+    ----------
+    data : list[dict]
+        Data product produced from processing respectively instrument.
+    data_table : dynamodb.Table
+        The DynamoDB table to insert or update the data.
+    instrument : str
+        The prefix for the product name.
+    """
+    # Get time range.
+    times = [item["met_in_utc"] for item in data]
+    min_time = min(times)
+    max_time = max(times)
+    logger.info(f"Processing {min_time} to {max_time}.")
+
+    science_data, hk_data = reformat_data(data)
+
+    # Query existing items.
+    response = data_table.query(
+        KeyConditionExpression=Key("instrument").eq(instrument)
+        & Key("time_utc").between(min_time, max_time)
+    )
+    hk_response = data_table.query(
+        KeyConditionExpression=Key("instrument").eq(f"{instrument}_hk")
+        & Key("time_utc").between(min_time, max_time)
+    )
+
+    existing_items = {item["time_utc"]: item for item in response.get("Items", [])}
+    existing_hk_items = {
+        item["time_utc"]: item for item in hk_response.get("Items", [])
+    }
```

Comment:
It looks like you are only looking at the time to see if there is an entry already in the table and ignoring it if so and not using the saved existing item. In that case I think you can simplify this to create a set of items to lookup in, maybe you can also have DynamoDB only filter to the time_utc for you by adding a FilterKey

```python
existing_science_times = set(item["time_utc"] for item in response.get("Items", []))

for data in science_data:
    if data["time_utc"] in existing_science_times:
        continue
    data_table.put_item(Item=data)
```